extension/chrome/settings/modules/contacts.ts

```diff
@@ -109,26 +110,44 @@ Catch.try(async () => {
 
   const actionProcessBulkImportTextInput = async () => {
     try {
-      const value = String($('#bulk_import .input_pubkey').val());
+      const value = Str.normalize(String($('#bulk_import .input_pubkey').val())).trim();
+      if (!value) {
+        await Ui.modal.warning('Please paste public key(s).');
+        return;
+      }
       const normalizedLongid = KeyImportUi.normalizeLongId(value);
-      let pubkey: string;
+      let pub: string;
       if (normalizedLongid) {
         const data = await Keyserver.lookupLongid(acctEmail, normalizedLongid);
         if (data.pubkey) {
-          pubkey = data.pubkey;
+          pub = data.pubkey;
         } else {
           await Ui.modal.warning('Could not find any Public Key in our public records that matches this fingerprint or longid');
           return;
         }
       } else {
-        pubkey = value;
+        pub = value;
       }
-      const replacedHtmlSafe = XssSafeFactory.replaceRenderableMsgBlocks(factory, pubkey);
-      if (replacedHtmlSafe && replacedHtmlSafe !== value) {
-        $('#bulk_import #processed').html(replacedHtmlSafe).css('display', 'block'); // xss-safe-factory
+      let { blocks } = Pgp.armor.detectBlocks(pub);
+      blocks = blocks.filter((b, i) => blocks.findIndex(f => f.content === b.content) === i); // remove duplicates
+      if (!blocks.length) {
+        await Ui.modal.warning('Could not find any new public keys.');
+      } else if (blocks.length === 1 && blocks[0].type === 'plainText') { // Show modal because users could make a mistake
+        await Ui.modal.warning('Incorrect public key. Please check and try again.');
+      } else { // Render Results
+        const container = $('#bulk_import #processed');
+        for (const block of blocks) {
+          if (block.type === 'publicKey') {
+            const replacedHtmlSafe = XssSafeFactory.replaceRenderableMsgBlocks(factory, String(block.content));
+            if (replacedHtmlSafe && replacedHtmlSafe !== value) {
+              container.append(replacedHtmlSafe); // xss-safe-factory
+            }
+          } else {
+            container.append(`<div class="bad">Skipping found '${Pgp.friendlyMsgBlockTypeName(block.type)}'</div>`); // xss-direct
```

Comment:
Just use sanitize append. I know it's safe here, and I know it's less efficient. But I will have less things to scrutinize during security reviews, and maybe it will be cheaper and simipler to do external security reviews, because they will also have fewer things to review. Thanks!